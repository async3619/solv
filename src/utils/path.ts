import * as os from "os";
import * as path from "path";
import * as fs from "fs-extra";
import { v4 as generateUUID } from "uuid";

const INSTANCE_ID = generateUUID();

export async function getTranspiledPath() {
    const targetDirectory = path.join(os.homedir(), ".solv");

    await fs.ensureDir(targetDirectory);
    return path.join(targetDirectory, `${INSTANCE_ID}.js`);
}
