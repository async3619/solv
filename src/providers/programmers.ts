import { parse as parseDOM } from "node-html-parser";

import { BaseProvider } from "./base";

import { findSiblingElement } from "../utils/findSiblingElement";
import { parseSampleTable } from "../utils/parseSampleTable";
import { Challenge } from "../utils/types";

export class ProgrammersProvider extends BaseProvider {
    public constructor() {
        super(false);
    }

    public checkUrl(url: string): boolean {
        return /(https?:\/\/)?(school\.)?programmers\.co\.kr\/learn\/courses\/([0-9]*?)\/lessons\/[0-9]([0-9]*?)$/.test(
            url,
        );
    }

    public serializeOutput(outputItem: any): string {
        return typeof outputItem === "string" ? `"${outputItem}"` : JSON.stringify(outputItem);
    }

    public beforeExecute(sourceCode: string): string {
        return [sourceCode, `console.info(solution(...global.input))`].join("\n\n");
    }

    public async retrieve(url: string): Promise<Challenge> {
        const targetUrl = new URL(url);
        targetUrl.searchParams.set("language", "javascript");

        const htmlCode = await this.fetch(targetUrl.toString());

        const document = parseDOM(htmlCode);
        const descriptionDOM = document.querySelector("#tour2 > div > div");
        const titleDOM = document.querySelector("#tab > li");
        if (!titleDOM?.textContent || !descriptionDOM?.textContent) {
            throw new Error("Failed to parse challenge information.");
        }

        const headers = document.querySelectorAll(".guide-section h5");
        const ioSampleHeaderRegex = /^입출력 ?예$/;
        const ioSampleHeaderDOM = headers.find(header => ioSampleHeaderRegex.test(header.textContent.trim()));
        if (!ioSampleHeaderDOM) {
            throw new Error("Failed to parse challenge input / output samples.");
        }

        const ioSampleTable = findSiblingElement(ioSampleHeaderDOM, node => node.tagName === "TABLE");
        if (!ioSampleTable) {
            throw new Error("Failed to find challenge output sample table.");
        }

        const initialCode = document
            .querySelector(`input[type="hidden"][name^="initial_code_"]`)
            ?.getAttribute("value");

        if (!initialCode) {
            throw new Error("Failed to parse initial code information.");
        }

        const id = targetUrl.pathname.split("/").at(-1);
        if (!id) {
            throw new Error("Failed to retrieve challenge id.");
        }

        return {
            ...parseSampleTable(ioSampleTable),
            id,
            title: titleDOM.textContent.trim(),
            description: descriptionDOM.textContent.trim(),
            provider: this,
            initialCode,
        };
    }
}
