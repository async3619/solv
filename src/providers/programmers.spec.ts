import fetch from "node-fetch";
import * as fs from "fs-extra";
import * as path from "path";

import { ProgrammersProvider } from "./programmers";

describe("Programmers Provider", () => {
    it("should get valid id from url", function () {
        const provider = new ProgrammersProvider();
        expect(provider.getId("https://school.programmers.co.kr/learn/courses/30/lessons/12937")).toBe(12937);
    });

    it("should determine valid url", () => {
        const provider = new ProgrammersProvider();
        expect(provider.checkUrl("https://school.programmers.co.kr/learn/courses/30/lessons/12937")).toBeTruthy();
    });

    it("should returns correct name", () => {
        const provider = new ProgrammersProvider();
        expect(provider.getName()).toBe("Programmers");
    });

    it("should determine invalid url", () => {
        const provider = new ProgrammersProvider();

        expect(provider.checkUrl("https://school.programmers.co.kr/learn/courses/30/lessons/")).toBeFalsy();
        expect(provider.checkUrl("https://school.programmers.co.kr/learn/lessons/12937")).toBeFalsy();
        expect(provider.checkUrl("https://school.programmers.co.kr/lessons/12937")).toBeFalsy();
    });

    it("should serialize output value properly", function () {
        const provider = new ProgrammersProvider();
        expect(provider.serializeOutput(123)).toBe("123");
        expect(provider.serializeOutput("123")).toBe(`"123"`);
    });

    it("should modify source code before run properly", () => {
        const provider = new ProgrammersProvider();
        expect(provider.beforeExecute("")).toMatchSnapshot();
    });

    it("should fail to parse challenge information from url when its corrupted", async () => {
        jest.mock("node-fetch");

        const provider = new ProgrammersProvider();
        const runTestWithMockedResponse = async (
            filePath: string,
            message: string,
            url = "https://school.programmers.co.kr/learn/courses/30/lessons/12937",
        ) => {
            const originalFetch = fetch;
            (fetch as any) = jest.fn().mockResolvedValue({
                ok: true,
                text() {
                    return fs.readFile(path.join(__dirname, filePath)).then(res => res.toString());
                },
            });

            await expect(provider.retrieve(url)).rejects.toThrow(message);

            (fetch as any) = originalFetch;
        };

        const messages = [
            "Failed to parse challenge input / output samples.",
            "Failed to find challenge output sample table.",
            "Failed to parse initial code information.",
            "Failed to retrieve challenge id.",
            "Failed to parse challenge information.",
            "Failed to parse challenge information.",
        ];

        for (let i = 0; i < messages.length; i++) {
            await runTestWithMockedResponse(
                `./__snapshots__/programmers.${i}.html`,
                messages[i],
                "https://school.programmers.co.kr/learn/courses/30/lessons/",
            );
        }
    });

    it("should parse challenge information from url properly", async () => {
        jest.dontMock("node-fetch");

        const provider = new ProgrammersProvider();
        const challenge = await provider.retrieve("https://school.programmers.co.kr/learn/courses/30/lessons/12937");

        expect(challenge).toMatchSnapshot();
    });
});
