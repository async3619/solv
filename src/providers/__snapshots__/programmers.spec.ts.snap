// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Programmers Provider should modify source code before run properly 1`] = `
"

console.info(solution(...global.input))"
`;

exports[`Programmers Provider should parse challenge information from url properly 1`] = `
{
  "description": "정수 num이 짝수일 경우 "Even"을 반환하고 홀수인 경우 "Odd"를 반환하는 함수, solution을 완성해주세요.

제한 조건


num은 int 범위의 정수입니다.
0은 짝수입니다.


입출력 예

        
num
return


        
3
"Odd"


4
"Even"",
  "id": "12937",
  "initialCode": "function solution(num) {
    var answer = '';
    return answer;
}",
  "input": [
    [
      "3",
    ],
    [
      "4",
    ],
  ],
  "output": [
    ""Odd"",
    ""Even"",
  ],
  "provider": ProgrammersProvider {
    "needJs": false,
  },
  "title": "짝수와 홀수",
}
`;
